docs/querying/dart.md

```diff
@@ -0,0 +1,116 @@
+
+You can query batch or realtime datasources with Dart.
+
+## Enable Dart
+
+To enable Dart, add the following line to your `broker/runtime.properties` and `historical/runtime.properties` files:
+
+```
+druid.msq.dart.enabled = true
+```
+
+### Configure resource consumption
+
+You can configure the Broker and the Historical to tune Dart's resource consumption.
+
+For Brokers, you can set the following configs:
+
+- `druid.msq.dart.controller.concurrentQueries`: The maximum number of query controllers that can run concurrently on that Broker. Additional controllers are queued. Defaults to 1.
+- `druid.msq.dart.query.context.targetPartitionsPerWorker`: The number of partitions per worker to create during a shuffle. Set this to the number of available threads on workers to fully take advantage of multi-threaded processing of shuffled data.
+
+For Historicals, you can set the following configs:
+
+- `druid.msq.dart.worker.concurrentQueries`: The maximum number of query workers that can run concurrently on that Historical. Default is equal to the number of merge buffers because each query needs one merge buffer. Ideally, this should be equal to or larger than the sum of the `concurrentQueries` setting on your Brokers.
+- `druid.msq.dart.worker.heapFraction`: The maximum amount of heap available for use across all Dart queries as a decimal. The default is 0.35, 35% of heap.
+
+
+## Run a Dart query
+
+Once enabled, you can use Dart in the Druid console or the SQL query API to issue queries.
+
+### Druid console
+
+In the **Query** view, select **Engine: SQL (Dart)** from the engine selector menu.
+
+### API
+
+Dart uses the SQL endpoint `/druid/v2/sql`. To use Dart, include the query context parameter `engine` and set it to `msq-dart`:
+
+<Tabs>
+  <TabItem value="SET" label="SET" default>
+    
+  ```sql
+  curl --location 'http://HOST:PORT/druid/v2/sql' \
+--header 'Content-Type: application/json' \
+--data '{
+  "query":   "SET engine = 'msq-dart';\nSELECT\n  user,\n  commentLength,COUNT(*) AS \"COUNT\" FROM wikipedia \nGROUP BY 1, 2 \nORDER BY 2 DESC",
+  ...
+  ...
+}'
+  ```
+
+  </TabItem>
+  <TabItem value="context_block" label="Context block">
+    
+  ```sql
+  curl --location 'http://HOST:PORT/druid/v2/sql' \
+  --header 'Content-Type: application/json' \
+  --data '{
+  "query":   "SELECT\n  user,\n  commentLength,COUNT(*) AS \"COUNT\" FROM wikipedia \nGROUP BY 1, 2 \nORDER BY 2 DESC",
+
+  ...
+  ...
+  "context": {
+    "engine":"msq-dart"
+    ...
+  }
+  }'
+  ```
+
+  </TabItem>
+  </Tabs>
+
+Dart supports many of the same [query context parameters as the MSQ task engine](../multi-stage-query/reference.md#context-parameters).
+
+  ## Known issues and limitations
```

Comment:
Some current known issues and limitations that come to mind for me:

- Dart doesn't verify that `druid.msq.dart.controller.concurrentQueries` is set properly, that's up to the admin. If set too high then queries can get stuck on each other.
- Dart does not use the query cache.
- Dart does not implement query prioritization or lanes.
- Dart (like MSQ in general) does not implement `useApproximateTopN`.
- Dart cannot be used with JDBC. The `engine` parameter is ignored.
- https://github.com/apache/druid/pull/18336 can in some cases lead to `NoClassDefFoundError` for `NilStageOutputReader`